File: question/neetcode150/binary_search/koko-eating-bananas.py
from typing import List
import math

class Solution:
    def minEatingSpeed(self, a: List[int], h: int) -> int:
        n = max(a)
        i = n

        while n:
            while i - n >= 0:
                total = sum(math.ceil(l/i) for l in a)
                if total <= h:
                    i-=n
            n//=2

        return i

File: question/neetcode150/binary_search/koko-eating-bananas.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool ok(vector<int>& a, int k, int h){
        int ans = 0;
        for(auto n:a){
            ans += n/k + (n%k!=0);
            if (ans > h){
                return false;
            }
        }
        return true;
    }

    int minEatingSpeed(vector<int>& a, int h) {
        int n = a.size();
        int k = *max_element(a.begin(), a.end());

        for (int i=k;i>0;i/=2){
            while(k-i>0 and ok(a,k-i,h)){
                k-=i;
            }
        }

        return k;
    }
};